src/Cake.Common/Build/BuildSystem.cs

```diff
@@ -114,6 +115,12 @@ public sealed class BuildSystem
             GoCD = goCDProvider;
             GitLabCI = gitlabCIProvider;
             TFBuild = tfBuildProvider;
+
+            Provider = typeof(BuildSystem)
```

Comment:
Instead of using Reflection, Linq, Substrings and SingleOrDefault (which could generate exceptions in constructor if you get more than one property), it would probably be more clear to just use an enum and the booleans. Roughly something like

 ```csharp
[Flags]
public enum BuildProvider
{
    Local = 0,
    AppVeyor = 1,
    TeamCity = 2,
    MyGet = 4,
    Bamboo = 8,
    ContinuaCI = 16,
    Jenkins = 32,
    Bitrise = 64,
    TravisCI = 128,
    BitbucketPipelines = 256,
    GoCD = 512,
    GitLabCI = 1024,
    AzurePipelines = 2048,
    AzurePipelinesHosted = 4096 
}
```
a used by a provider property
```csharp
public BuildProvider Provider { get; }
```

assigned in constructor something like this
```csharp
Provider =  (AppVeyor.IsRunningOnAppVeyor ? BuildProvider.AppVeyor : BuildProvider.Local)
        | (TeamCity.IsRunningOnTeamCity ? BuildProvider.TeamCity : BuildProvider.Local)
        | (MyGet.IsRunningOnMyGet ? BuildProvider.MyGet : BuildProvider.Local)
        | (Bamboo.IsRunningOnBamboo ? BuildProvider.Bamboo : BuildProvider.Local)
        | (ContinuaCI.IsRunningOnContinuaCI ? BuildProvider.ContinuaCI : BuildProvider.Local)
        | (Jenkins.IsRunningOnJenkins ? BuildProvider.Jenkins : BuildProvider.Local)
        | (Bitrise.IsRunningOnBitrise ? BuildProvider.Bitrise : BuildProvider.Local)
        | (TravisCI.IsRunningOnTravisCI ? BuildProvider.TravisCI : BuildProvider.Local)
        | (BitbucketPipelines.IsRunningOnBitbucketPipelines ? BuildProvider.BitbucketPipelines : BuildProvider.Local)
        | (GoCD.IsRunningOnGoCD ? BuildProvider.GoCD : BuildProvider.Local)
        | (GitLabCI.IsRunningOnGitLabCI ? BuildProvider.GitLabCI : BuildProvider.Local)
        | (TFBuild.IsRunningOnTFS ? BuildProvider.AzurePipelines : BuildProvider.Local)
        | (TFBuild.IsRunningOnVSTS ? BuildProvider.AzurePipelinesHosted : BuildProvider.Local);

IsLocalBuild = Provider == BuildProvider.Local;
```

Then `IsLocalBuild` could be something like
```csharp
public bool IsLocalBuild { get; }
```

If for some reason you would run an AzurePipelines agent on i.e. AppVeyor 😎 then

```csharp
Information(
   new {
      BuildSystem.Provider,
      BuildSystem.IsLocalBuild
   }
   );
```
would output
```
{ Provider = AppVeyor, AzurePipelines, IsLocalBuild = False }
```
and not through an exception.

Reply:
An enum would be better, but flags doesn’t seem right and adds its own complexity. Surely we can assume the build providers, including local, are mutually exclusive? (That’s why I used `SingleOrDefault`.)

Also, looks like you’re agreeable to rename `VSTS` to `AzurePipelines` (for hosted)? Is there an equivalent of `TFS` though; i.e. is there on-premise version of Azure Pipelines?